runtime/near-vm-runner/src/logic/logic.rs

```diff
@@ -2518,7 +2524,7 @@ impl<'a> VMLogic<'a> {
         tracing::trace!(
             target = "io_tracer",
             storage_op = "read",
-            key = %near_fmt::Bytes(&key),
+            key = base64(&key),
```

Comment:
Yeah unfortunately this loses one of the main use cases of the `Bytes` smart pretty printing. (here in `read` but also for the other three changes in this file)

For example:

```
storage_read READ key='STATE' size=70 tn_db_reads=20 tn_mem_reads=0 shard_cache_miss=19 shard_cache_hit=2
```
This shows the SDK reading the magic key `"STATE"`, which `near_gmt::Bytes` detects as ASCII only and therefore prints in a easily interpretable way. With your change, it will be just another unreadable base64 encoded key.

I'd be sad if we lose that debugging feature. That said, it wouldn't break any of the tooling around IO trace, since parsers already handle base64.

To work around this, perhaps we can move tracing output to the `RuntimeExt` side? I'm pretty sure we don't need it on `MockedExternal`. And I believe all the information printed is also available a call layer deeper.

Reply:
Oh right, I was thinking no one ever read these logs, but I guess contract developers running the sandbox would actually be seeing them. So while it’d make downstream tools simpler too (because fewer corner cases, eg. I wonder how `'` in the keys are handled by the downstream tools as `Bytes` can output it), we do lose one user-facing benefit.

As for putting the tracing to `RuntimeExt`, that’d be a solution, but then we’d be hitting the same issue once we try to remove the deps from `runtime` too. WDYT about copy-pasting the `Bytes` and. `AbbrBytes` into near-vm-runner as a `helpers.rs` file? It’s only ever used for debug logging anyway, and it’s like 200loc, so hopefully it wouldn’t be too bad.

Or we could publish near-fmt as a freestanding crate on crates.io like finite-wasm, as it feels like it’s pretty isolated and feature-complete right now. Or we could try using the already-existing `display-bytes` crate, that seems like it could replicate the behavior of near-fmt pretty well.

WDYT?

Reply:
(I have a personal preference for using the `display_bytes` crate, because it’s just less code for us to maintain, but it doesn’t have the exact `AbbrBytes` behavior and I know some people are reluctant to new dependencies)

Reply:
Sounds like a great idea! Just to confirm how much I should validate that everything will still be fine with this change: production does not run with `io_trace` on, right?

Reply:
Oh. I had forgotten about it, but `tracing` needs its values to implement `tracing::Value`, which AFAICT cannot include a byte slice by default.

There are two ways to work around it. One would be to use `valuable`, but that’d most likely be a mess given that AFAIU it’d create a tree (from the tracing point of view) with one node per byte, and then the trace layer would have to rebuild the slice and do its thing.

The other one would be to keep the code as currently is, so running base64 all the time; and then to have the IO trace layer un-base64 when it can.

I’m partial for the second idea, as it gives a clear API (the value in tracing is a base64), while allowing for easy UX (because the UX-related part presents it as best as it can for the user).

WDYT?

Reply:
Got it! Then I’ve just pushed a commit with the changes we chatted about, hopefully this is now good to go :)

Reply:
Yes, looks good now, thanks!